### DISCUSSION TITLE
Having trouble to deploy on Heroku

### DISCUSSION THREAD

**Discussion by kevinka999** [adonisjs/core#1969] [Help]
Hello guys, I'm trying to deploy my API Adonis JS 5.0 on Heroku.
I'm using **Heroku Postgres** for my database.
I already saw a discussion about the same error, but didn't solve for me.

In my **_package.json_**, my script "start" is that
`ENV_SILENT=true node build/server.js`
And my **_Procfile_** is that
`release: ENV_SILENT=true node ace migration:run --force`
`web: ENV_SILENT=true npm start`

To be honest, I already got 3 differents troubles.
The first one is `Error: AdonisJS requires "@adonisjs/assembler" in order to run typescript source directly`

I already tried `npm install @adonisjs/assembler` and add `"@adonisjs/assembler": "^3.0.6"` at the depedencies on the **_package.json_**, but still getting that error.

The second one is a `Exception: E_MISSING_ENV_VALUE: Missing environment variable "HOST"`
But I already put all values necessary (that I think) for a postgres pg, the enviroment variables is:
`**DB_CONNECTION, PG_HOST, PG_PORT, PG_USER, PG_PASSWORD, PG_DB_NAME**`

And the last error I got (now I dont have it, because when I tried to solve, I got back to the beggining)
When tried to get something to the route, I got `(query) 'table' doesn't exist`

Well. I dont know where to find help about this.
If someone help me, I will be grateful :D

**Comment by iammrsea** [comment]
I'm not sure why you were setting `ENV_SILENT=true` before every command. The best way is to just set it once as environment variable.
You can do this from your local terminal if you have heroku cli locally installed, alternatively, you can set from your heroku dashboard under settings. 

For your problems, I would have taken the following steps if I were in your shoes:

1. Log in to your heroku account and select your app ( I'm assuming that you had already created an app).

2. Click the **settings tab** and scroll down to **Reveal Config Vars** and click it.

3. Add the following environment variables-```
HOST=0.0.0.0
APP_KEY=your app key (you can just copy the one in the .env file on your local machine)
DB_CONNECTION=pg
PG_HOST=localhost
PG_PORT=5432
PG_USER=postgres
PG_PASSWORD=password
ENV_SILENT=true
PG_DB_NAME=name```

**Note, PG host, port, user,password and name are just to make adonjs happy so that your app would build successfully because you wont't be connecting to your database with them. Hence, you can just fill them in with the same values you had in .env file during development.**

If you look closely on the environment variables on your heroku dashboard, you would see **DATABASE_URL** added automatically, that's what I normally use to connect to the postgres database.

That should be all you need for env variables.

### Coming back to your application
Go to the config folder, open **database.ts** file and replace the content with the code below:

``` js
import Env from '@ioc:Adonis/Core/Env'
import { OrmConfig } from '@ioc:Adonis/Lucid/Orm'
import { DatabaseConfig } from '@ioc:Adonis/Lucid/Database'
import  Application  from '@ioc:Adonis/Core/Application';

const databaseConfig: DatabaseConfig & { orm: Partial<OrmConfig> } = {
  connection: Env.get('DB_CONNECTION'),

  connections: {
    pg: {
      client: 'pg',
      connection: Application.inProduction ?
      Env.get('DATABASE_URL') + "?ssl=no-verify":
      {
        host: Env.get('PG_HOST'),
        port: Env.get('PG_PORT'),
        user: Env.get('PG_USER'),
        password: Env.get('PG_PASSWORD', ''),
        database: Env.get('PG_DB_NAME'),
      },
    
      healthCheck: Application.inDev,
			debug: Application.inDev,
    },

/*
I usually use this custom connection to run database migrations on remote database from my local machine. For instance,
node ace migration:run --connection=custom, this will run the migration against you remote database. In order to use this connection, you must set DATABASE_URL in .env file, you can get the value from your heroku dashboard, or by running this command on your terminal: heroku config:get DATABASE_URL --app=your_app_name
*/
    custom: {
      client: "pg",
      connection: Env.get("DATABASE_URL") + "?ssl=no-verify",
    },

  },

  orm: {
  },
}

export default databaseConfig
```
These are usually my workflow when I work with heroku, you don't really have to follow them. But if you do, redeploy your app again
and let me know if you still have any challenges

**Reply by kevinka999** [reply]
First, thanks for helping!

Now I can access my routes, but when routes try to access my database, I got the third error `"(table) does not exist"`.
And I still getting `Error: AdonisJS requires "@adonisjs/assembler" in order to run typescript source directly` in my logs (crying a lot here)

**Reply by kevinka999** [reply]
News. I tried to run my migrations manually by **_console_** using `heroku run bash` and I got the error `"AdonisJS requires "@adonisjs/assembler"`.
So I presume error is when I run migrations.

Someone have any ideia?

**Reply by kevinka999** [reply]
I got a solution!!!
To run migrations in prod, we need to run in the folder **/build**

So I create a new **.sh** file called `release.sh` and add the code bellow
`cd build`
`ENV_SILENT=true node ace migration:run --force`

and in my **Procfile** I changed to
`release: sh ./release.sh`

**Reply by cawa-93** [reply]
@iammrsea How validate Environment Variables in your case? Make all optional? Or required? 

```ts
Env.rules({
  DATABASE_URL: Env.schema.string({format: 'url'}),
  
  PG_HOST: Env.schema.string({format: 'host'}),
  PG_PORT: Env.schema.number(),
  PG_USER: Env.schema.string(),
  PG_PASSWORD: Env.schema.string.optional(),
  PG_DB_NAME: Env.schema.string(),
})
```

I tried use condition 

```ts
  ...(
    Env.get('DATABASE_URL', undefined) !== undefined
      ? {DATABASE_URL: Env.schema.string({format: 'url'})}
      : {
        PG_HOST: Env.schema.string({format: 'host'}),
        PG_PORT: Env.schema.number(),
        PG_USER: Env.schema.string(),
        PG_PASSWORD: Env.schema.string.optional(),
        PG_DB_NAME: Env.schema.string(),
      }
  ),
```
  
But this lead to TS compile error: 
```
  config/app.ts:154:3 - error TS2322: Type 'unknown' is not assignable to type 'string'.

154   level: Env.get('LOG_LEVEL', 'info'),

      ~~~~~
```
